### PR TITLE
Move sparse tests to TestOptimRenewed

### LINKED ISSUE #103322: Disabling ALL TestOptim on the dynamo config
### 🐛 Describe the bug

https://github.com/pytorch/pytorch/pull/102640 introduced flakiness into the dynamo optim tests.

There was a followup PR https://github.com/pytorch/pytorch/pull/103066 to try to disable the tests in attempt to restore health, but that did not sufficiently cover all forms of flakiness as there were further reports of flaky tests after. 

This issue tracks the fact that we currently disable ALL test_optim tests in the dynamo shard, which is probably undesirable, but is a necessary stopgap to restore CI health. The alternative is reverting https://github.com/pytorch/pytorch/pull/102640 and its forward fixes, like https://github.com/pytorch/pytorch/pull/103121. 

### Versions

main

cc @vincentqb @jbschlosser @albanD @crcrpar @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @ezyang @gchanan @zou3519 @anijain2305 @Xia-Weiwen @ipiszy

### PR DESCRIPTION
**Pull request by janeyx99**
This is the last of the old TestOptim! With this change, everything will be migrated to use OptimizerInfo. Our sparse support is...well, sparse, and the tests try to best encapsulate which configs actually work. Note that support_sparse is actually just supports sparse grads...we don't test sparse params. 

1. This PR fixes a bug in Adagrad multi_tensor with maximize by passing the correct value of maximize (vs False everytime) when sparse values are present.

2. This PR does improve coverage. There used to only be 2 configs each, and now we have the following configs for:

Adagrad:
```
python test/test_optim.py -k test_rosenbrock_sparse_with_lrsched_False_Adagrad
/home/janeyx/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
{'maximize': True, 'lr': 0.1}
{'initial_accumulator_value': 0.1, 'lr': 0.1}    <--- this and above are CPU
.{'foreach': False, 'lr': 0.1}
{'foreach': True, 'lr': 0.1}
{'maximize': True, 'foreach': False, 'lr': 0.1}
{'maximize': True, 'foreach': True, 'lr': 0.1}
{'initial_accumulator_value': 0.1, 'foreach': False, 'lr': 0.1}
{'initial_accumulator_value': 0.1, 'foreach': True, 'lr': 0.1}
.
----------------------------------------------------------------------
Ran 2 tests in 227.744s

OK
```

SGD
```
(pytorch-3.10) [janeyx@devgpu023.odn1 /data/users/janeyx/pytorch (bff23193)]$ python test/test_optim.py -k test_rosenbrock_sparse_with_lrsched_False_SGD
/home/janeyx/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
{'dampening': 0.5, 'lr': 0.0048}
.{'foreach': False, 'lr': 0.0048}
{'foreach': True, 'lr': 0.0048}
{'dampening': 0.5, 'foreach': False, 'lr': 0.0048}
{'dampening': 0.5, 'foreach': True, 'lr': 0.0048}
.
----------------------------------------------------------------------
Ran 2 tests in 112.801s

OK
```

SparseAdam
```
(pytorch-3.10) [janeyx@devgpu023.odn1 /data/users/janeyx/pytorch (bff23193)]$ python test/test_optim.py -k test_rosenbrock_sparse_with_lrsched_False_Sparse
/home/janeyx/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
{'maximize': True, 'lr': 0.04}
.{'maximize': True, 'lr': 0.04}
.
----------------------------------------------------------------------
Ran 2 tests in 35.113s

OK
```

Fixes #103322. A side quest in this migration was to re-enable and track dynamo issues as they trigger on the optim tests, which will be complete from this PR. New tests may add more things to track in dynamo, but there is now an established system for doing so, and dynamo is either enabled or a bug is tracked for every migrated test in TestOptimRenewed.

Next steps:
Remove the hyperparameter constraints in common_optimizer.py defined by metadata_for_sparse (other than LR, which seems handpicked for the tests to actually pass). Doing this requires adding more sparse functionality.

Add more tests!

Maybe add more optimizers!

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123146
* #123139
* #123134

### CODE REVIEW

File: test/test_optim.py [hunk 1]
```diff
@@ -34,6 +34,12 @@ def rosenbrock(tensor):
     return (1 - x) ** 2 + 100 * (y - x**2) ** 2
 
 
+def drosenbrock(tensor):
```
Comment (albanD) [inline test/test_optim.py:37]: What does the b mean here?

File: test/test_optim.py [hunk 2]
```diff
@@ -34,6 +34,12 @@ def rosenbrock(tensor):
...
+    assert tensor.size() == torch.Size([2]), f"Requires tensor with 2 scalars but got {tensor.size()}"
+    x, y = tensor
+    return torch.tensor((-400 * x * (y - x**2) - 2 * (1 - x), 200 * (y - x**2)))
```
Comment (albanD) [inline test/test_optim.py:40]: Why re-wrap in a tensor() here? I think torch.stack() would be better

File: test/test_optim.py [hunk 3]
```diff
@@ -249,6 +255,129 @@ def closure(optim, w, b, i):
...
+
+            def get_grad(param, sparse_grad, w):
+                grad = drosenbrock(param)
+                # NB: We torture test the optimizer by returning an
+                # uncoalesced sparse tensor
+
+                # Depending on w, provide only the x or y gradient
+                if sparse_grad:
+                    if w:
+                        i = torch.LongTensor([[0, 0]])
```
Comment (albanD) [inline test/test_optim.py:319]:
```suggestion
                        i = torch.tensor([[0, 0]], dtype=torch.int64)
```
Same below

File: torch/optim/adagrad.py [hunk 4]
```diff
@@ -338,7 +338,7 @@ def _multi_tensor_adagrad(
                 lr_decay=lr_decay,
                 eps=eps,
                 has_sparse_grad=True,
-                maximize=False,
+                maximize=maximize,
```
Comment (albanD) [inline torch/optim/adagrad.py:341]: Good catch!

**Review by albanD — APPROVED** [review]
Thanks!

File: test/test_optim.py (hunk 1 above)
Comment (janeyx99) [reply test/test_optim.py:37]: Do you mean d? It's the derivative

**Pull request closed without merging**